crates/mongodb-connector/src/api_type_conversions/query_request.rs

```diff
@@ -203,53 +358,90 @@ fn v3_to_v2_order_by_element(
-// v2-compatible). This will be done in an upcoming PR.
-fn v3_to_v2_target_path(path: Vec<v3::PathElement>) -> Result<Vec<String>, ConversionError> {
-    fn is_expression_non_empty(expression: &v3::Expression) -> bool {
-        match expression {
-            v3::Expression::And { expressions } => !expressions.is_empty(),
-            v3::Expression::Or { expressions } => !expressions.is_empty(),
-            _ => true,
-        }
-    }
-    if path
-        .iter()
-        .any(|path_element| match &path_element.predicate {
-            Some(pred) => is_expression_non_empty(pred),
-            None => false,
-        })
-    {
-        Err(ConversionError::NotImplemented(
-            "The MongoDB connector does not currently support predicates on references through relations",
-        ))
-    } else {
-        Ok(path.into_iter().map(|elem| elem.relationship).collect())
+fn v3_to_v2_target_path(
+    context: &QueryContext,
+    collection_relationships: &BTreeMap<String, v3::Relationship>,
+    root_collection_object_type: &WithNameRef<schema::ObjectType>,
+    path: Vec<v3::PathElement>
+) -> Result<(Vec<String>, HashMap<String, v2::OrderByRelation>), ConversionError> {
+    let mut v2_path = vec![];
+    let v2_relations = v3_to_v2_target_path_step::<Vec<_>>(context, collection_relationships, root_collection_object_type, path.into_iter(), &mut v2_path)?;
+    Ok((v2_path, v2_relations))
+}
+
+fn v3_to_v2_target_path_step<T : IntoIterator<Item = v3::PathElement>>(
+    context: &QueryContext,
+    collection_relationships: &BTreeMap<String, v3::Relationship>,
+    root_collection_object_type: &WithNameRef<schema::ObjectType>,
+    mut path_iter: T::IntoIter, 
+    v2_path: &mut Vec<String>
+) -> Result<HashMap<String, v2::OrderByRelation>, ConversionError> {
+    let mut v2_relations = HashMap::new();
+
+    if let Some(path_element) = path_iter.next() {
+        v2_path.push(path_element.relationship.clone());
+
+        let where_expr = path_element
+            .predicate
+            .map(|expression| {
+                let v3_relationship = lookup_relationship(collection_relationships, &path_element.relationship)?;
+                let target_collection = context.find_collection(&v3_relationship.target_collection)?;
+                let target_object_type = context.find_object_type(&target_collection.r#type)?;
+                let v2_expression = v3_to_v2_expression(context, collection_relationships, root_collection_object_type, &target_object_type, *expression)?;
+                Ok(Box::new(v2_expression))
+            })
+            .transpose()?;
+
+        let subrelations = v3_to_v2_target_path_step::<T>(context, collection_relationships, root_collection_object_type, path_iter, v2_path)?;
+        
+        v2_relations.insert(
+            path_element.relationship, 
+            v2::OrderByRelation {
+                r#where: where_expr,
+                subrelations,
+            }
+        );
```

Comment:
Ok so you're hoisting the order_by predicates. Good thinking. I forget if we still need to do something similar for comparison targets. I'm not sure if my TODO comment still applies to that case.

Reply:
No, with comparison targets in predicates, and predicates from the permissions system (which is what these order by predicates are for) should already been inlined into the predicate body.